src/dbnode/client/replicated_session.go

```diff
@@ -0,0 +1,301 @@
+	replicateError       tally.Counter
+}
+
+func newReplicatedSessionMetrics(scope tally.Scope) replicatedSessionMetrics {
+	return replicatedSessionMetrics{
+		replicateExecuted:    scope.Counter("replicate.executed"),
+		replicateNotExecuted: scope.Counter("replicate.not-executed"),
+		replicateError:       scope.Counter("replicate.error"),
+	}
+}
+
+// Ensure replicatedSession implements the clientSession interface.
+var _ clientSession = (*replicatedSession)(nil)
+
+type replicatedSessionOption func(*replicatedSession)
+
+func withNewSessionFn(fn newSessionFn) replicatedSessionOption {
+	return func(session *replicatedSession) {
+		session.newSessionFn = fn
+	}
+}
+
+func newReplicatedSession(opts ReplicatedOptions, options ...replicatedSessionOption) (clientSession, error) {
+	// TODO(srobb): Replace with PooledWorkerPool once it has a GoIfAvailable method
+	workerPool := m3sync.NewWorkerPool(maxReplicationConcurrency)
+	workerPool.Init()
+
+	scope := opts.Options().InstrumentOptions().MetricsScope()
+
+	session := replicatedSession{
+		newSessionFn: newSession,
+		workerPool:   workerPool,
+		scope:        scope,
+		log:          opts.Options().InstrumentOptions().Logger(),
+		metrics:      newReplicatedSessionMetrics(scope),
+		outCh:        make(chan error),
+	}
+
+	// Apply options
+	for _, option := range options {
+		option(&session)
+	}
+
+	if err := session.setSession(opts.Options()); err != nil {
+		return nil, err
+	}
+	if err := session.setAsyncSessions(opts.OptionsForAsyncClusters()); err != nil {
+		return nil, err
+	}
+
+	return &session, nil
+}
+
+type writeFunc func(Session) error
+
+func (s *replicatedSession) setSession(opts Options) error {
+	if opts.TopologyInitializer() == nil {
+		return nil
+	}
+
+	session, err := s.newSessionFn(opts)
+	if err != nil {
+		return err
+	}
+	s.session = session
+	return nil
+}
+
+func (s *replicatedSession) setAsyncSessions(opts []Options) error {
+	sessions := make([]clientSession, 0, len(opts))
+	for _, oo := range opts {
+		session, err := s.newSessionFn(oo)
+		if err != nil {
+			return err
+		}
+		sessions = append(sessions, session)
+	}
+	s.asyncSessions = sessions
+	return nil
+}
+
+func (s replicatedSession) replicate(fn writeFunc) error {
+	for _, asyncSession := range s.asyncSessions {
+		asyncSession := asyncSession // capture var
+		if s.workerPool.GoIfAvailable(func() {
+			err := fn(asyncSession)
+			if err != nil {
+				s.metrics.replicateError.Inc(1)
+				s.log.Error("could not replicate write: %v", zap.Error(err))
+			}
+			s.outCh <- err
+		}) {
+			s.metrics.replicateExecuted.Inc(1)
+		} else {
+			s.metrics.replicateNotExecuted.Inc(1)
+		}
+	}
+	return fn(s.session)
+}
+
+// Write value to the database for an ID
+func (s replicatedSession) Write(namespace, id ident.ID, t time.Time, value float64, unit xtime.Unit, annotation []byte) error {
+	return s.replicate(func(session Session) error {
+		return session.Write(namespace, id, t, value, unit, annotation)
+	})
```

Comment:
Using lambdas will unfortunately create a heap allocation per call (to capture the args). Maybe it's possible to pass the args as a struct and switch if/else on tagged or not?